Automattic/sensei
# Update enrolled courses queries to use learner term

Fixes #2863

#### Changes proposed in this Pull Request:

* Centralizes enrolled course querying in `Sensei_Learner`. This could live in `Sensei_Course` or `Sensei_Course_Enrolment_Manager`, but both seemed a bit bloated and I thought this seemed okay.
* Updated query on _Course Component_ widget for `My Active Courses` and `My Completed Courses`.
* Updated query on _My Courses_ page (`[sensei_user_courses]`).
* Updated query on _Learner Profile_ pages (`https://example.com/learner/{username}`). 
* Deprecated using `usercourses` for `$type` in `Sensei_Course::course_query` and `Sensei_Course::get_archive_query_args`.
* Added a notice on _My Courses_ on the _All Courses_ tab when no courses are found. We could consolidate these methods (or have one call the other with a argument).
* Ensures user enrolment is calculated before querying.

In this PR, I _might_ have fixed the legacy shortcodes, but I didn't test them.

Additional notes:
- _My Courses_ ignores course order and instead sorts by title. _Learner Profiles_ uses course order if set and `date` if not. Should we standardize this? If so, to what?
- The `My Courses Pagination` setting has had an effect on Learner Profile pages. Should we rename this to be more general?
- I think `Sensei_Course::course_query` and `Sensei_Course::get_archive_query_args` are just used in the legacy shortcodes class. We could deprecate them completely. **Edit: I've deprecated them now**
- "Calculating enrolment" is much more lightweight than Sensei 2/WCPC 1. @gkaragia's `\Sensei_Course_Enrolment_Manager::recalculate_enrolments` method, which is called before any of the above queries, only actually calculates enrolment if necessary. Before, we were doing all those heavy queries every time we loaded _My Courses_. The benefit of this is we can also ensure course enrolment is up to date in these other places.

#### Testing instructions:

- Enroll a student in several courses. Complete some of them.
- Make sure there are some courses a student isn't enrolled in.
- Enable Learner Profiles if not already enabled.
- In settings under _Courses_, set the `My Courses Pagination` to a small number (less than the number of courses a student is enrolled in).
- Visit _Learner Profile_ page. Make sure listings show up correctly. Check pagination.
- Visit _My Courses_ page. Make sure listings show up correctly. Check pagination.
- Bonus: Start from pre-3.0.0 database with the above conditions set (disable WCPC for now as its enrolment is incomplete). Move to this branch with all WordPress pages closed in browser and immediately load the _My Courses_ page. It should calculate enrolment and show the correct courses.

#### Deprecations
- `Sensei_Course::filter_my_courses` 
- `Sensei_Course::course_query`
- `Sensei_Course::get_archive_query_args`
